module/VuFindSearch/src/VuFindSearch/Backend/Feature/ConnectorCacheTrait.php

```diff
@@ -0,0 +1,127 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * @category VuFind
+ * @package  Search
+ * @author   Ere Maijala <ere.maijala@helsinki.fi>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org Main Site
+ */
+namespace VuFindSearch\Backend\Feature;
+
+use Laminas\Cache\Storage\StorageInterface;
+use Laminas\Http\Client as HttpClient;
+
+/**
+ * Caching support trait for connectors.
+ *
+ * @category VuFind
+ * @package  Search
+ * @author   Ere Maijala <ere.maijala@helsinki.fi>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org Main Site
+ */
+trait ConnectorCacheTrait
+{
+    /**
+     * Request cache
+     *
+     * @var StorageInterface
+     */
+    protected $cache = null;
+
+    /**
+     * Set the cache storage
+     *
+     * @param StorageInterface $cache Cache
+     *
+     * @return void
+     */
+    public function setCache(StorageInterface $cache)
+    {
+        $this->cache = $cache;
+    }
+
+    /**
+     * Create a cache key from client's request state
+     *
+     * @param HttpClient $client HTTP Client
+     *
+     * @return string
+     */
+    public function getCacheKey(HttpClient $client): string
+    {
+        return md5($client->getRequest()->toString());
+    }
+
+    /**
+     * Get a request from cache if available
+     *
+     * @param string $key Cache key
+     *
+     * @return mixed
+     */
+    public function getCachedData(string $key)
+    {
+        try {
+            if ($result = $this->cache->getItem($key)) {
+                $this->debug('Returning cached results');
+                return $result;
+            }
+        } catch (\Exception $ex) {
+            $this->logWarning('Cache getItem failed: ' . $ex->getMessage());
+        }
+        return null;
+    }
+
+    /**
+     * Cache response data.
+     *
+     * @param string $key      Cache entry key
+     * @param mixed  $response Response to be cached
+     *
+     * @return void
+     */
+    protected function putCachedData(string $key, $response): void
+    {
+        try {
+            $this->cache->setItem($key, $response);
+        } catch (\Laminas\Cache\Exception\RuntimeException $ex) {
+            // Try to determine if caching failed due to response size and log the
+            // case accordingly. Unfortunately Laminas Cache does not translate
+            // exceptions to any common error codes, so we must check codes and/or
+            // message for adapter-specific values.
+            // 'ITEM TOO BIG' is the message from the Memcached adapter
+            // and comes directly from libmemcached.
+            if ($ex->getMessage() === 'ITEM TOO BIG') {
+                $this->debug('Cache setItem failed: ' . $ex->getMessage());
```

Comment:
Is the idea that we only want to use debug level for TOO BIG because it's a common/expected problem, and we want unexpected errors to get higher logging priority? If that's the case, I wonder if we should also consider using a different message for the debug case -- maybe something more specific -- to make this more clear.

Reply:
Yes, it can be a common occurrence, and should not pollute the logs unless debug level is enabled.

Worth mentioning, since we discussed the error message handling during the [initial implementation](https://github.com/vufind-org/vufind/pull/1854#discussion_r576839049): While the error code from memcached is now available, I didn't feel like it's any better to compare the code with a seemingly magic value of `37` (available as a constant in Memcached extension, but would need to check the presence of the extension first). If you think it would be better, I can change it.

Reply:
I updated the debug message test.

Reply:
I think it's worth changing to use the error code, since that seems less likely to change in the future. Since checking for the constant is extra work, maybe the best thing to do is to compare against the magic number but put a comment indicating that the number is also defined as a constant, but the constant may not be available due to the optional nature of the PHP extension.

Reply:
Now changed. And that's a good thing. Now that I verified the functionality, the exception message is a bit confusingly "ITEM TOO BIG; SUCCESS".